File: task2.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    static List<string> FindUnique(List<string> v)
    {
        return v.Distinct().ToList();
    }

    static string Format(string s)
    {
        var result = new System.Text.StringBuilder();
        bool seenAt = false;
        bool seenStar = false;
        foreach (char c in s)
        {
            if (c == '@') seenAt = true;
            if (!seenAt)
            {
                if (c == '*') { seenStar = true; continue; }
                if (seenStar) continue;
                if (c == '.') continue;
            }
            result.Append(c);
        }
        return result.ToString();
    }

    static bool IsAllowed(string s)
    {
        char[] notAllowedChars = { '&', '=', '+', '-', '_', '<', '>', ',', '\'' };
        if (string.IsNullOrEmpty(s) || s[0] == '.') return false;
        if (!s.Contains('@')) return false;

        int atIndex = s.IndexOf('@');
        if (atIndex < 6 || atIndex > 30) return false;

        foreach (char c in s.Substring(0, atIndex))
        {
            if (notAllowedChars.Contains(c)) return false;
        }

        if (s.LastIndexOf('.') <= atIndex || !s.Contains('.')) return false;
        return true;
    }

    static void Main()
    {
        var v = new List<string> { ".555555@.", "1111.111@mail.ru", "11111111@.mailru", "1111111@.mailru", "11111111@.mailru" };
        var filtered = v.Where(IsAllowed).Select(Format).ToList();
        if (filtered.Count == 0) Console.WriteLine("No unique emails");
        else Console.WriteLine(FindUnique(filtered).Count);
    }
}

File: task2.kt
fun findUnique(v: List<String>): List<String> {
    val unique = mutableListOf<String>()
    for (i in v) {
        var isUnique = true
        for (j in unique) {
            if (j == i) {
                isUnique = false
                break
            }
        }
        if (isUnique) unique.add(i)
    }
    return unique
}

fun format(s: String): String {
    var result = s
    var iter = 0
    while (iter < result.length && result[iter] != '@') {
        if (result[iter] == '*') {
            while (iter < result.length && result[iter] != '@') {
                result = result.removeRange(iter, iter + 1)
            }
        } else if (result[iter] == '.') {
            result = result.removeRange(iter, iter + 1)
        } else {
            iter++
        }
    }
    return result
}

fun isAllowed(s: String): Boolean {
    val notAllowedChars = charArrayOf('&', '=', '+', '-', '_', '<', '>', ',', '\'')
    if (s.isEmpty() || s[0] == '.') return false
    if (!s.contains('@')) return false
    
    val atIndex = s.indexOf('@')
    if (atIndex < 6 || atIndex > 30) return false
    
    for (i in 0 until atIndex) {
        if (s[i] in notAllowedChars) return false
    }
    
    if (s.lastIndexOf('.') <= atIndex || !s.contains('.')) return false
    return true
}

fun main() {
    val v = listOf(".555555@.", "1111.111@mail.ru", "11111111@.mailru", "1111111@.mailru", "11111111@.mailru")
    val filtered = v.filter { isAllowed(it) }.map { format(it) }
    if (filtered.isEmpty()) println("No unique emails")
    else println(findUnique(filtered).size)
}